File: leetcode/238.product-of-array-except-self.py
# @leet start
class Solution:
    def productExceptSelf(self, nums: list[int]) -> list[int]:
        # Initialize res[i] to store the product of the last n - 1 - i
        # numbers
        res = [1] * len(nums)

        for i in range(len(nums) - 2, -1, -1):
            res[i] = nums[i + 1] * res[i + 1]

        # Products except self
        prefix_prod = nums[0]

        for i in range(1, len(nums)):
            res[i] *= prefix_prod
            prefix_prod *= nums[i]

        return res


# @leet end

File: leetcode/238.product-of-array-except-self.ts
// @leet start
function productExceptSelf(nums: number[]): number[] {
  const result: number[] = new Array(nums.length);

  let prod = 1;

  for (let i = 0; i < nums.length; i++) {
    result[i] = prod;
    prod *= nums[i];
  }

  prod = 1;

  for (let i = nums.length - 1; i >= 0; i--) {
    result[i] *= prod;
    prod *= nums[i];
  }

  return result;
}
// @leet end
